### PR TITLE
Run and collect benchmarks

### PR DESCRIPTION
**Pull request by liyuqian**
This will start to collect engine benchmarks in the flutter-cirrus
Datastore for all post-submit commits.

We're using this to test how https://github.com/liyuqian/metrics_center
works. Once it's stablized, we should move metrics_center into
flutter/packages, and migrate data to flutter-infra Datastore.

Related issue: https://github.com/flutter/flutter/issues/37434

### CODE REVIEW

**Comment by liyuqian** [comment]
@cbracken @godofredoc @digiter @keyonghan : I hope that we can start collecting benchmarks for flutter/engine before the end of 2019 :)

**Comment by liyuqian** [comment]
@goderbauer @digiter @keyonghan : I uploaded this first for review because I want this to test https://github.com/liyuqian/metrics_center first. After the testing is satisfied, I'll ask you to review https://github.com/liyuqian/metrics_center , and either merge it into cocoon or flutter/packages.

File: testing/benchmark/bin/parse_and_send.dart [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+Future<List<FlutterEngineMetricPoint>> _parse(String jsonFileName) async {
+  final GitDir gitDir = await GitDir.fromExisting('../../');
+  // Somehow gitDir.currentBranch() doesn't work in Cirrus with "fatal: 'HEAD' -
+  // not a valid ref". Therefore, we use "git log" to get the revision manually.
+  final ProcessResult logResult = await gitDir
+      .runCommand(<String>['log', '--pretty=format:%H]', '-n', '1']);
+  if (logResult.exitCode != 0) {
+    throw 'Unexpected exit code ${logResult.exitCode}';
+  }
+  final String gitRevision = logResult.stdout.toString();
```
Comment (digiter) [inline testing/benchmark/bin/parse_and_send.dart]: Nit: encapsulate the logic of getting git revision into a sub function.

File: testing/benchmark/bin/parse_and_send.dart [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:git/git.dart';
+import 'package:metrics_center/flutter.dart';
+
+const String kNameKey = 'name';
+const String kTimeUnitKey = 'time_unit';
+
+const List<String> kSkippedSubResults = <String>[
```
Comment (digiter) [inline testing/benchmark/bin/parse_and_send.dart]: Why skip these sub results?

File: testing/benchmark/bin/parse_and_send.dart [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          FlutterEngineMetricPoint(name, value, gitRevision,
+              moreTags: <String, String>{kSubResultKey: subResult}
+                ..addAll(context)
+                ..addAll(subResult.endsWith('time')
+                    ? timeUnitMap
+                    : <String, String>{})),
+        );
+      }
+    }
+  }
```
Comment (digiter) [inline testing/benchmark/bin/parse_and_send.dart]: Nit: I would also put parsing an item into a sub function.

File: testing/benchmark/bin/parse_and_send.dart [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+  return points;
+}
+
+Future<void> main(List<String> args) async {
+  if (args.length != 1) {
```
Comment (digiter) [inline testing/benchmark/bin/parse_and_send.dart]: Consider using ArgParser()..addFlag, https://dart.dev/tutorials/server/cmdline

File: testing/benchmark/bin/parse_and_send.dart [hunk 5]
```diff
@@ -0,0 +1,84 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (keyonghan) [inline testing/benchmark/bin/parse_and_send.dart]: 2019?

File: testing/benchmark/bin/parse_and_send.dart [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:git/git.dart';
+import 'package:metrics_center/flutter.dart';
+
+const String kNameKey = 'name';
+const String kTimeUnitKey = 'time_unit';
+
```
Comment (keyonghan) [inline testing/benchmark/bin/parse_and_send.dart]: Nit: remove empty line.

File: testing/benchmark/bin/parse_and_send.dart [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+      jsonDecode(File(jsonFileName).readAsStringSync());
+
+  final Map<String, dynamic> rawContext = jsonResult['context'];
+  final Map<String, String> context = rawContext.map<String, String>(
+    (String k, dynamic v) => MapEntry<String, String>(k, v.toString()),
+  );
+
+  final List<FlutterEngineMetricPoint> points = <FlutterEngineMetricPoint>[];
+  for (Map<String, dynamic> item in jsonResult['benchmarks']) {
+    final String name = item[kNameKey];
```
Comment (keyonghan) [inline testing/benchmark/bin/parse_and_send.dart]: Nit: remove empty lines between variables, and add one before the for loop?

File: testing/benchmark/bin/parse_and_send.dart [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  return points;
+}
+
+Future<void> main(List<String> args) async {
+  if (args.length != 1) {
+    throw 'Must have one argument: <benchmark_json_file>';
+  }
+  final List<FlutterEngineMetricPoint> points = await _parse(args[0]);
```
Comment (keyonghan) [inline testing/benchmark/bin/parse_and_send.dart]: One check to make sure the file(args[0]) exists?

File: testing/benchmark/bin/parse_and_send.dart [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  final FlutterDestination destination =
```
Comment (keyonghan) [inline testing/benchmark/bin/parse_and_send.dart]: Maybe good to document which DB this link points to.

File: testing/benchmark/test/parse_and_send_test.dart [hunk 10]
```diff
@@ -0,0 +1,20 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (keyonghan) [inline testing/benchmark/test/parse_and_send_test.dart]: same here. 2019?

File: testing/benchmark/pubspec.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+name: flutter_engine_benchmark
+
+dependencies:
+  git: any
+  metrics_center:
+    # TODO(liyuqian): once metrics_center is properly reviewed, add it to
+    # flutter/packages, publish on pub.dev, and use the published package here.
+    git: https://github.com/liyuqian/metrics_center.git
```
Comment (digiter) [inline testing/benchmark/pubspec.yaml]: What does metrics center do and why not integrating parse_and_send.dart into metrics center?

File: testing/benchmark/test/parse_and_send_test.dart [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+// found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:test/test.dart';
+
+void main() {
+  test('parse_and_send with example json does not crash.', () async {
+    final String testCred =
+        File('secret/test_gcp_credentials.json').readAsStringSync();
```
Comment (digiter) [inline testing/benchmark/test/parse_and_send_test.dart]: This "secret" directory is git-ignored, then how could another one run this test?

File: testing/benchmark/bin/parse_and_send.dart (hunk 5 above)
Comment (keyonghan) [reply testing/benchmark/bin/parse_and_send.dart]: Oh, I see.

**Comment by liyuqian** [comment]
Rerun successful at https://ci.chromium.org/p/flutter/builders/try/Linux%20Host%20Engine/284 . Merging.

**Pull request merged**